app/helpers/payment_details_helper.rb

```diff
@@ -98,4 +99,8 @@ def country_code_for(country_name)
 
     merchant_countries.find { |country| country[0] == country_name }&.dig(1)
   end
+
+  def ip_address
```

Comment:
Well, this is a simple way to get the IP address of the client... 
It is called on the form rendering, not on submitting the Braintree form, I guess this should be OK. It might not work correctly with some network infras.

An alternative could be calling some 3rd party service like https://www.ipify.org/ from JavaScript, but I think that's not ideal either.

Reply:
Seems like this is the canonical approach that should work with a proper infra configuration. We may perform a safety check to verify that a public IP address is returned and refuse to process the request if so. For example if it returns 127.0.0.1 or 10.2.3.23

But that's it. We can't account for broken network infrastructure.

Reply:
Well, then it will break in local testing, because locally we do get `127.0.0.1` :sweat_smile: 

And also, I think we shouldn't fail. If anything, 3DS should fail to authorize the card with an invalid address. And in case no 3DS is used, then we shouldn't even care about the IP address value.

Reply:
agreed, but good to add a log line that the IP doesn't look right

Reply:
If this code is in the server side, is this IP address used in the automated charging process? Or is it sent to the browser so it knows which IP send to the PG in as 3DS field?

If the 3DS authorization is entirely happening between the browser and the PG without our intervention, and the PG can see which IP the request is coming from, why would they need us to send any IP at all?

Reply:
> If the 3DS authorization is entirely happening between the browser and the PG without our intervention, and the PG can see which IP the request is coming from, why would they need us to send any IP at all?

That is the unanswered question. It could be just to know what IP was used for one part of the process and for the other part of the process. As an additional fraud detection data point. Or it might be the case that we don't actually need to... Somebody has to go carefully over the requirements and or get an answer from PG support.